### DISCUSSION TITLE
How to Create a Client Keyring Using Go-Ceph?

### DISCUSSION THREAD

**Discussion by ThingOfNull** [ceph/go-ceph#949] [Q&A]
I'm interested in using Go-Ceph to create a client keyring for Ceph-Fuse. Previously, I used the following command to create a keyring:

```ceph auth get-or-create client.abcdef  mon 'allow r' mds 'allow rw path=/abcdef' osd 'allow *'```

Is it possible to achieve this with Go-Ceph? If not, is there a feature that allows remote creation of keyrings? (I don't want to log in to the server to create a keyring every time.)

thanks.

**Comment by phlogistonjohn** [comment]
Generally, `ceph` command line tools are passed to the cluster via a `mon_command` function. The go-ceph version is at [rados.MonCommand](https://pkg.go.dev/github.com/ceph/go-ceph/rados#Conn.MonCommand). However, it's not as simple as just passing the CLI command to that function as the CLI is converted into a JSON structure first. It's not particularly complex either.

Because we support other projects that want convenient APIs to access these ceph commands we have a number of "admin" apis that handle the JSON serialization/de-serialization into Go data structures. Examples: [cephfs/admin](https://pkg.go.dev/github.com/ceph/go-ceph/cephfs/admin), [common/admin/nfs](https://pkg.go.dev/github.com/ceph/go-ceph@v0.25.0/common/admin/nfs), [common/admin/manager](https://pkg.go.dev/github.com/ceph/go-ceph@v0.25.0/common/admin/manager).  You could look at the code that implements these packages and see how the JSON objects are created before being passed to the MonCommand or MgrCommand funcs. 

I would also be open to the idea of adding a `common/admin/auth` sub-package to go-ceph. File a feature request issue if you'd like. If you're particularly interested in the idea we'd also welcome contributions in that vein. :-)

See also: https://github.com/ceph/go-ceph/blob/master/docs/hints.md#cant-find-the-api-you-want
